File: src/main/kotlin/common2/TreeNode.java
package common2;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    public static TreeNode buildTree(Integer... values) {
        if (values == null || values.length == 0 || values[0] == null) {
            return null; // return empty tree if no values or first value is null
        }

        TreeNode root = new TreeNode(values[0]); // first value is the root
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int i = 1;

        while (!queue.isEmpty() && i < values.length) {
            TreeNode current = queue.poll();

            // Assign left child if present
            if (i < values.length && values[i] != null) {
                current.left = new TreeNode(values[i]);
                queue.offer(current.left);
            }
            i++;

            // Assign right child if present
            if (i < values.length && values[i] != null) {
                current.right = new TreeNode(values[i]);
                queue.offer(current.right);
            }
            i++;
        }

        return root;
    }

    // Method to print the tree structure
    void printTree() {
        int depth = maxDepth(this);
        int maxWidth = (1 << depth) - 1;

        String[][] levels = new String[depth][maxWidth];
        for (String[] level : levels) {
            Arrays.fill(level, " ");
        }

        fillLevels(levels, this, 0, 0, maxWidth);

        for (String[] level : levels) {
            System.out.println(String.join("", level));
        }
    }

    // Helper method to calculate the maximum depth of the tree
    private int maxDepth(TreeNode node) {
        return node == null ? 0 : 1 + Math.max(maxDepth(node.left), maxDepth(node.right));
    }

    // Recursive function to fill levels of the tree
    private void fillLevels(String[][] levels, TreeNode node, int depth, int left, int right) {
        if (node == null) return;
        int mid = left + (right - left) / 2;
        levels[depth][mid] = String.valueOf(node.val);

        fillLevels(levels, node.left, depth + 1, left, mid);
        fillLevels(levels, node.right, depth + 1, mid + 1, right);
    }

}

File: src/main/kotlin/common/TreeNode.kt
package common

import java.util.LinkedList
import java.util.Queue

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null

    companion object {
        fun fromList(vararg levelOrder: Int?): TreeNode? {
            if (levelOrder.isEmpty() || levelOrder[0] == null) return null

            val root = TreeNode(levelOrder[0]!!)
            val queue: Queue<TreeNode> = LinkedList()
            queue.add(root)

            var i = 1
            while (i < levelOrder.size) {
                val current = queue.poll()

                if (i < levelOrder.size && levelOrder[i] != null) {
                    current.left = TreeNode(levelOrder[i]!!)
                    queue.add(current.left)
                }
                i++

                if (i < levelOrder.size && levelOrder[i] != null) {
                    current.right = TreeNode(levelOrder[i]!!)
                    queue.add(current.right)
                }
                i++
            }

            return root
        }
    }

    fun printTree() {
        fun maxDepth(node: TreeNode?): Int = node?.let { 1 + maxOf(maxDepth(it.left), maxDepth(it.right)) } ?: 0

        val depth = maxDepth(this)
        val maxWidth = (1 shl depth) - 1

        val levels = Array(depth) { arrayOfNulls<String>(maxWidth) }
        fillLevels(levels, this, 0, 0, maxWidth)

        for (level in levels) {
            println(level.joinToString("") { it ?: " ".repeat(maxWidth / level.size) })
        }
    }

    fun fillLevels(levels: Array<Array<String?>>, node: TreeNode?, depth: Int, left: Int, right: Int) {
        if (node == null) return
        val mid = left + (right - left) / 2
        levels[depth][mid] = node.`val`.toString()

        fillLevels(levels, node.left, depth + 1, left, mid)
        fillLevels(levels, node.right, depth + 1, mid + 1, right)
    }
}
